runtime/gc_structs/FlattenedContiguousArrayIterator.hpp

```diff
@@ -0,0 +1,144 @@
+	fj9object_t *_basePtr; /**< pointer to the first array slot, where element 0 is */
+	fj9object_t *_scanPtr; /**< pointer to the current array element's first slot */
+	fj9object_t *_endPtr; /**< pointer to the array slot where the iteration will terminate */
+	J9Class *_elementClass; /**< Pointer to class of the elements in the flattened array */
+protected:
+	OMR_VM *_omrVM;
+public:
+	MMINLINE GC_SlotObject *nextSlot()
+	{
+		/* If no more object slots to scan, returns NULL */
+		GC_SlotObject *result = NULL;
+		if (_scanPtr < _endPtr) {
+			result = _mixedObjectIterator.nextSlot();
+			if (NULL == result) {
+				_scanPtr = (fj9object_t *)(_elementStride + (uintptr_t)_scanPtr);
+				if (_scanPtr < _endPtr) {
+					/* mixedObjectIterator reached end of element. Move the iterator to the beginning of the next element */
+					_mixedObjectIterator.initialize(_omrVM, _elementClass, _scanPtr);
+					result = _mixedObjectIterator.nextSlot();
+				}
+			}
+		}
+		return result;
+	}
+
+	MMINLINE void initialize(J9Object *objectPtr)
+	{
+		MM_GCExtensionsBase *extensions = MM_GCExtensionsBase::getExtensions(_omrVM);
+		GC_ArrayObjectModel *arrayObjectModel = &(extensions->indexableObjectModel);
+		J9Class *clazzPtr = J9GC_J9OBJECT_CLAZZ(objectPtr, extensions);
+
+		_elementStride = J9ARRAYCLASS_GET_STRIDE(clazzPtr);
+		_basePtr = (fomrobject_t *)arrayObjectModel->getDataPointerForContiguous(_arrayPtr);
+		_scanPtr = _basePtr;
+		_endPtr = (fomrobject_t *)((uintptr_t)_basePtr + (arrayObjectModel->getSizeInElements(_arrayPtr) * _elementStride));
+		_elementClass = ((J9ArrayClass *) clazzPtr)->componentType;
+
+		if (_scanPtr < _endPtr) {
+			_mixedObjectIterator.initialize(_omrVM, _elementClass, _scanPtr);
+		}
+	}
+
+	/**
+	 * Gets the current slot's array index.
+	 * @return slot number (or zero based array index) of the last call to nextSlot.
+	 */
+	MMINLINE UDATA getIndex()
+	{
+		return ((uintptr_t)_scanPtr - (uintptr_t)_basePtr) / _elementStride;
+	}
+
+	/**
+	 * Sets the current slot's array index
+	 * @param[in] index index to set scan index to
+	 */
+	MMINLINE void setIndex(UDATA index) {
+		_scanPtr = (fj9object_t *)((uintptr_t)_basePtr + (index * _elementStride));
+	}
+
+	/**
+	 * Restores the iterator state into this class
+	 * @param[in] objectIteratorState partially scanned object iterator state
+	 */
+	MMINLINE void restore(GC_ObjectIteratorState *objectIteratorState)
+	{
+		_scanPtr = objectIteratorState->_scanPtr;
+		_endPtr = objectIteratorState->_endPtr;
+	}
+
+	/**
+	 * Saves the partially scanned state of this class
+	 * @param[in] objectIteratorState where to store the state
+	 */
+	MMINLINE void save(GC_ObjectIteratorState *objectIteratorState)
+	{
+		objectIteratorState->_scanPtr = _scanPtr;
+		objectIteratorState->_endPtr  = _endPtr;
+	}
+
+	MMINLINE J9Object *getObject()
+	{
+		return (J9Object *)_arrayPtr;
+	}
+
+	GC_FlattenedContiguousArrayIterator(OMR_VM *omrVM, J9Object *objectPtr)
+		: _arrayPtr((omrarrayptr_t)objectPtr)
```

Comment:
The declared type of `_arrayPtr` is `J9IndexableObject *`, so I believe it is better to cast to this type